### ISSUE TITLE
Script API: expose pathfinding to the script

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1979]
### Problem

Currently pathfinding in AGS is used strictly for the built-in behavior. If a user would like to script their own movement they would have to use a workaround, like a invisible character moving around, and attaching real object to it.

### IMPORTANT

Below are original API proposals, but discussion that followed brought up several issues with these. Please refer to the comments below, or the [summarizing comment](https://github.com/adventuregamestudio/ags/issues/1979#issuecomment-1510406790), to see what we came up to. This ticket has to be updated after we have a better API draft.

### Proposal

Implement a script function that calculates path between starting and ending position. For example:
```
int[] Room.FindPath(int sx, int sy, int ex, int ey);
```
```
Point[] Room.FindPath(int sx, int sy, int ex, int ey);
```
There are two alternatives here, one returning an array of ints, another an array of managed Points.
Each may have pros and cons. Points are easier syntactically, but will require to allocate N managed objects, and will take bit more memory because of how managed objects are stored (unfortunately at this moment ags script does not support returning regular structs by values, neither dynamic arrays of regular structs for some reasons).
In case of ints, the x & y coordinates would be stored sequentially (x0, y0, x1, y1, and so forth).

In AGS4 the new compiler already supports .Length property in arrays, but if we want to also support compiler without this feature, then we might add one final element that signifies end of data: e.g. null pointer in case of `Point[]` array, and some obvious invalid value in case of `int[]` (like INT32_MIN).

Implementation would call current pathfinder (see `find_route`). Something to consider: I don't remember exactly, but I think that pathfinder uses "movement speed" parameter when calculating moves. In any case, we do not need step-by-step points for this, only node points.

See also, as mildly relevant: #1965

### Alternate proposal

After writing this ticket, I realized there may be an alternative to `int[]`/`Point[]` array. That is - a `MovePath` struct, which might look like:
```
builtin managed struct MovePath {
    import readonly attribute Point Pos[];
    // OR
    import readonly attribute int X[];
    import readonly attribute int Y[];
    //
    import readonly attribute int Length;
};
```
This is somewhat more complicated, but there are couple of advantages I can see:
* this struct may be easily expanded if necessary;
* we may use same struct for exposing Character's and Object's move paths (like - `Character.MovePath` and `Object.MovePath` properties).
* there's already an internal MoveList object, so it may have certain sense to make a script's "frontend" for the MoveList.

**Comment by ericoporto** [comment]
Hey, I have a plugin for this if you want to experiment with the API: https://github.com/ericoporto/agspathfinder

**Comment by ivan-mogilko** [comment]
> Hey, I have a plugin for this if you want to experiment with the API: https://github.com/ericoporto/agspathfinder

Hmm, so the things to consider are:
1. Running pathfinding on a custom bitmap rather than only the room's walkable area.
2. Knowing the current position along the path, at least between which two nodes the mover is.

With the MovePath struct the p2 may be solved with `int CurrentPos`, for example.

**Comment by ericoporto** [comment]
Just expanding a bit in your p1 crude draft... I don't remember the exact details of the pathfinding, but I expect it has to use the bitmap mask to make nodes.

Assuming this is a thing, when using `FindPathInRoom` it will use the already generated nodes, supposing they exist.

Now, `FindPathOnMask` would then generate these nodes again on each call (we don't know if the sprite changed in between), if this was not in a static struct, it could generate the nodes once, and hold the state until asked again.

Back to `FindPathInRoom`, it magically pics up blocking objects and characters with their rules - I think except the player character. But in `FindPathOnMask` you handle manually with mask drawing any necessary blocking object you want to exist.

Now, on a tangential, `FindPathOnMask` has no concept of walkable masks, and is 32-bit (or the game bit depth), so my guess is 0 is wall, and any other value is walkable.

**Comment by ivan-mogilko** [comment]
> Now, FindPathOnMask would then generate these nodes again on each call (we don't know if the sprite changed in between), if this was not in a static struct, it could generate the nodes once, and hold the state until asked again.

Right... so we would require some persistent object that is generated from mask, and used further for pathfinding.

> Back to FindPathInRoom, it magically pics up blocking objects and characters with their rules - I think except the player character.

No, there's no magic there. The engine has a copy of walkable mask and cuts holes in it for blocking objects. This is done before pathfinder.

> Now, on a tangential, FindPathOnMask has no concept of walkable masks, and is 32-bit (or the game bit depth), so my guess is 0 is wall, and any other value is walkable.

That's a good point, I was assuming it will be using 8-bit masks, but there's currently no way to *create* 8-bit masks explicitly in script, only getting room masks.
There are two potential solutions here:
* we add a method for creating dynamicsprite with the 8-bit color depth for mask purposes;
* additionally pass a color-to-area table into the pathfinder for 32-bit images (this is something I was forced to do in my racing game).

**Comment by ivan-mogilko** [comment]
To summarize:
There has to be a persistent object containing navigation data, called like `PathGrid` or `NavGrid` etc.
This object hypothetically may be constructed in different ways, but we stick to mask for now.
It's used to find paths from one point to another; the path is returned either as an array of points, or another special object (allows easier extension).
AGS Room struct can return its own "path grid", which corresponds to the actual walkable mask it uses.

**Comment by ivan-mogilko** [comment]
I've been reviewing this conversation lately, and have more thoughts on this.

In my opinion, it's clear that there's a minimal "path" type, which is - an array of Points, and this is enough for the basic low-level pathfinding API. Anything else would complicate things or represent a extended API. I suppose that we may have both simple API and more complex API later, provided there will be methods of constructing more complex "extended path" objects out of a simple "path" (array of points), and vice versa.

If we look at the pathfinder implementation in the engine right now, we will notice that it does 2 jobs:
* builds a list of checkpoints, that is - points at which moving object's direction changes;
* builds a list of move vectors, which depend on moving speeds, and let "tween" an object between each two checkpoints.

I think that for the most basic API the second task may be skipped, as users may like to decide on the speed and movement themselves. After all, the velocity is trivial to calculate, it's the path itself which is a problem.
Similarly remembering "current position" is an extended information that may be easily added in user script. E.g. if you wrap an array of Points into a struct that also has "current pos".
So, we might have to split the existing pathfinder into 2 separate jobs, one of which finds the path and returns a list of points, and another generates "MoveList" using that path and movement speed parameters.

I know that we discussed a "pathfinder grid" object earlier, but believe that we might postpone adding that, because that will be a separate object that requires its own treatment.

---

My proposal for the first step would be again to have a most basic function in the Room struct:
```
Point[] Room.FindPath(int sx, int sy, int ex, int ey);
```
And functions in Character and Object that let accept this array:
```
Character.WalkPath(Point[], ...);
Object.MovePath(Point[], ...);
```

**Comment by ericoporto** [comment]
That minimal proposal would be nice, I think you can instead for the parameters use like `x0`, `y0`, `x1`, `y1` or `start_x`, `start_y`, `target_x`, `target_y`. I think that target is a good name that it's a target, but you may get a different final point (if that point is not walkable).

**Comment by ivan-mogilko** [comment]
Ok, I actually had a sudden realization about "Pathfinder object", maybe I will come with something soon.
But I must refactor existing route finder code first, because currently it is buried under "helper" functions, and depends on static global data.

**Comment by ivan-mogilko** [comment]
Basic api is resolved by #2441
Anything else may be added on top of that later.
